kubernetes/websocket/kube_exec.c

```diff
@@ -0,0 +1,122 @@
+{
+    int command_string_length = 0;
+    const char *p = original_command_string;
+    char escaped_string[ESCAPED_STRING_BUFFER_SIZE];
+    while (*p) {
+        if (' ' == *p) {
+            command_string_length += strlen(COMMAND_PREFIX);
+            if (command_string_length >= url_command_string_buffer_size) {
+                fprintf(stderr, "%s: The length of command string exceeds the buffer size.", __func__);
+                return -1;
+            }
+            strncat(command_string_in_url, COMMAND_PREFIX, strlen(COMMAND_PREFIX));
+        } else {
+            memset(escaped_string, 0, sizeof(escaped_string));
+            escape_character_in_url(escaped_string, ESCAPED_STRING_BUFFER_SIZE, *p);
+            command_string_length += strlen(escaped_string);
+            if (command_string_length >= url_command_string_buffer_size) {
+                fprintf(stderr, "%s: The length of command string exceeds the buffer size.", __func__);
+                return -1;
+            }
+            strncat(command_string_in_url, escaped_string, strlen(escaped_string));
+        }
+        p++;
+    }
+    return 0;
+}
+
+int kube_exec(wsclient_t * wsc, const char *namespace_, const char *pod_name, const char *container_name, int stdin, int stdout, int tty, const char *command)
+{
+    int rc = 0;
+
+    const char *ws_path_template = "/api/v1/namespaces/%s/pods/%s/exec?stdin=%s&stdout=%s&tty=%s&command=%s";
+
+    /* 
+     * e.g. 
+     * /api/v1/namespaces/default/pods/pod1/exec?stdin=true&stdout=true&tty=false&command=ls&command=/&container=container1
+     * /api/v1/namespaces/default/pods/pod1/exec?stdin=true&stdout=true&tty=true&command=bash&container=container1
+     */
+
+    char command_string_in_url[URL_COMMAND_STRING_BUFFER_SIZE];
+    memset(command_string_in_url, 0, sizeof(command_string_in_url));
+    rc = assemble_command_in_url(command_string_in_url, URL_COMMAND_STRING_BUFFER_SIZE, command);
+    if (rc != 0) {
+        return rc;
+    }
+
+    char ws_path[WS_PATH_BUFFER_SIZE];
+    memset(ws_path, 0, sizeof(ws_path));
+    snprintf(ws_path, WS_PATH_BUFFER_SIZE, ws_path_template,
+             namespace_, pod_name, 1 == stdin ? TRUE_STRING : FALSE_STRING, 1 == stdout ? TRUE_STRING : FALSE_STRING, 1 == tty ? TRUE_STRING : FALSE_STRING, command_string_in_url);
+
+    if (container_name && container_name[0] != '\0') {
+        int length_of_ws_path_with_container = strlen(ws_path) + strlen(CONTAINER_PREFIX) + strlen(container_name);
+        if (length_of_ws_path_with_container >= WS_PATH_BUFFER_SIZE) {
+            fprintf(stderr, "%s: Error: The length of the websocket path exceeds the buffer size.", __func__);
+            return -1;
+        }
+        strncat(ws_path, CONTAINER_PREFIX, strlen(CONTAINER_PREFIX));
+        strncat(ws_path, container_name, strlen(container_name));
+    }
+
+    if (wsc->log_mask & LLL_USER) {
+        printf("%s: ws_path=%s\n", __func__, ws_path);
+    }
+    wsc->path = strdup(ws_path);
```

Comment:
Is this memory allocation necessary?

Reply:
No, it's not.
But in wsclient_free
```c
void wsclient_free(wsclient_t * wsc)
{
...
    if (wsc->path) {
        free(wsc->path);
        wsc->path = NULL;
    }
...
}

```

So we need ensure `wsc->path == NULL` to avoid coredump

maybe I can change as below ?
```c
wsc->path = ws_path;
···
rc = wsclient_run(wsc, mode);
wsc->path = NULL;
reurn rc;
```

Reply:
I should have been more clear: Perhaps the ws client shouldn't own the `path` variable, so you don't need to ever delete it.